enhancements/machine-health-checking.md

```diff
@@ -0,0 +1,204 @@
+---
+title: machine-health-checking
+authors:
+  - @bison
+  - @ingvagabund
+  - @enxebre
+  - @cynepco3hahue
+reviewers:
+  - @derekwaynecarr
+  - @michaelgugino
+  - @beekhof
+  - @mrunalp
+approvers:
+  - @enxebre
+  - @bison
+  - @derekwaynecarr
+  - @michaelgugino
+
+creation-date: 2019-09-09
+last-updated: 2019-09-09
+status: implementable
+see-also:
+replaces:
+superseded-by:
+---
+
+# Machine health checking
+
+## Release Signoff Checklist
+
+ - [x] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift/docs
+
+
+## Summary
+
+Enable opt in for automated health checking and remediation of unhealhty nodes backed by machines.
+
+## Motivation
+
+Reduce administrative overhead to run a cluster and ability to respond to failures of machines.
+
+### Goals
+
+- Providing a default remediation strategy for public clouds - Machine deletion.
+
+- Allowing to define different unhealthy criterias for different pools of machines.
+
+- Allowing new and custom remediation strategies developed outside the machine-healthcheck controller, e.g baremetal reboot.
+
+- Providing administrator-defined short-circuiting of automated remediation.
+
+
+### Non-Goals
+
+- Coordinating individual remediation systems. This must be driven by each system however they see fit, e.g. by annotating affected nodes with state and transition time.
+
+- Recording long-term stable history of all health-check failures or remediations.
+
+## Proposal
+
+- As a customer I only care about my apps availability so I want my cluster infrastructure to be self healing and the nodes to be remediated automatically.
+
+- As a cluster admin I want to define my unhealthy criteria as a list of node conditions https://github.com/kubernetes/node-problem-detector.
+
+- As a cluster admin I want to define different unhealthy criteria for pools of machines targeted for different workloads.
+
+- As a cluster admin I want to define different remediation strategies for different cloud environments. E.g deletion or reboot.
+
+- As a cluster admin I want to skip remediation to optionally prevent automated remediation if some threshold of nodes are unhealthy, because this situation may require more careful intervention. E.g a large number of nodes in a single zone are down because it's most likely a networking issue.
```

Comment:
I think an additional useful limitation would be NN% of nodes per hour. If we're somehow interpreting something benign as a problem, I would want this additional circuit-breaker.

Reply:
Is there any reason not to have the exact same behavior as node controller for zone failures?

Reply:
> Is there any reason not to have the exact same behavior as node controller for zone failures?

What is that behavior?

Reply:
The node lifecycle controller has the view of all the nodes of the cluster per zone at any time, which is not necessarily the case here as this would allow targeted pools of machines. We provide the building blocks though so as a user you can set it up to have that similar behaviour if that's what you want. I'd be in favour of introducing built-in smarter magic in a later release.